src/sparse.c

```diff
@@ -0,0 +1,706 @@
+ */
+
+#include "sparse.h"
+#include "attrcache.h"
+#include "git2/sparse.h"
+#include "config.h"
+#include "filebuf.h"
+#include "index.h"
+
+static bool sparse_lookup_in_rules(
+        int *checkout,
+        git_attr_file *file,
+        git_attr_path *path)
+{
+	size_t j;
+	git_attr_fnmatch *match;
+	
+	git_vector_rforeach(&file->rules, j, match) {
+		if (match->flags & GIT_ATTR_FNMATCH_DIRECTORY &&
+				path->is_dir == GIT_DIR_FLAG_FALSE)
+			continue;
+		if (git_attr_fnmatch__match(match, path)) {
+			*checkout = ((match->flags & GIT_ATTR_FNMATCH_NEGATIVE) == 0) ?
+			GIT_SPARSE_CHECKOUT : GIT_SPARSE_NOCHECKOUT;
+			return true;
+		}
+	}
+	
+	return false;
+}
+
+static int parse_sparse_file(
+        git_repository *repo,
+        git_attr_file *attrs,
+        const char *data,
+        bool allow_macros)
+{
+	int error = 0;
+	int ignore_case = false;
+	const char *scan = data;
+	git_attr_fnmatch *match = NULL;
+	GIT_UNUSED(allow_macros);
+	
+	if (git_repository__configmap_lookup(&ignore_case, repo, GIT_CONFIGMAP_IGNORECASE) < 0)
+		git_error_clear();
+	
+	if (git_mutex_lock(&attrs->lock)) {
+		git_error_set(GIT_ERROR_OS, "failed to lock sparse-checkout file");
+		return -1;
+	}
+	
+	while (!error && *scan) {
+		int valid_rule = 1;
+		
+		if (!match && !(match = git__calloc(1, sizeof(*match)))) {
+			error = -1;
+			break;
+		}
+		
+		match->flags =
+		GIT_ATTR_FNMATCH_ALLOWSPACE | GIT_ATTR_FNMATCH_ALLOWNEG;
+		
+		if (!(error = git_attr_fnmatch__parse(
+			match, &attrs->pool, NULL, &scan)))
+		{
+			match->flags |= GIT_ATTR_FNMATCH_IGNORE;
+			
+			if (ignore_case)
+				match->flags |= GIT_ATTR_FNMATCH_ICASE;
+			
+			scan = git__next_line(scan);
+			
+			/*
+			 * If a negative match doesn't actually do anything,
+			 * throw it away. As we cannot always verify whether a
+			 * rule containing wildcards negates another rule, we
+			 * do not optimize away these rules, though.
+			 * */
+			if (match->flags & GIT_ATTR_FNMATCH_NEGATIVE
+					&& !(match->flags & GIT_ATTR_FNMATCH_HASWILD))
+				error = git_attr__does_negate_rule(&valid_rule, &attrs->rules, match);
+			
+			if (!error && valid_rule)
+				error = git_vector_insert(&attrs->rules, match);
+		}
+		
+		if (error != 0 || !valid_rule) {
+			match->pattern = NULL;
+			
+			if (error == GIT_ENOTFOUND)
+				error = 0;
+		} else {
+			match = NULL; /* vector now "owns" the match */
+		}
+	}
+	
+	git_mutex_unlock(&attrs->lock);
+	git__free(match);
+	
+	return error;
+}
+
+int git_sparse_attr_file__init_(
+		int *file_exists,
+        git_repository *repo,
+        git_sparse *sparse)
+{
+    int error = 0;
+	git_str infopath = GIT_STR_INIT;
+    const char *filename = GIT_SPARSE_CHECKOUT_FILE;
+	git_attr_file_source source = { GIT_ATTR_FILE_SOURCE_FILE, git_str_cstr(&infopath), filename, NULL };
+	git_str filepath = GIT_STR_INIT;
+
+	if ((error = git_str_joinpath(&infopath, repo->gitdir, "info")) < 0) {
+		if (error != GIT_ENOTFOUND)
+			goto done;
+		error = 0;
+	}
```

Comment:
Would like to have used something like `git_repository__item_path` here, but it would always get the info path for the root repository, not the worktrees. Keen to know if there's a nicer way to do this.